### PR TITLE
Fiducial Follow new commands and status operation

### PR DESCRIPTION
**Pull request by mjstn**
Here is the greatly enhanced fiducial follow that is meant to be backward compatible by default.
I am going to ask Alan to pull and run it at least in default mode for 2nd verification it works like before.

Rohan and David are basically FYI as this should NOT be something Rohan should spend his school time on unless desired.   Jim:  Sorry but you are the official 'reviewer' if you could please.  I don't feel there is a rush.

The real value here is of course to supply a flexible tool to our users that can really get them going with as simple as possible interface to understand and code to due to the supplied client example making it rather simple to operate and make magni move about.

### CODE REVIEW

File: custom_messages/msg/FollowerCommand.msg
Comment (jim-v) [inline custom_messages/msg/FollowerCommand.msg:6]: These field names are not very descriptive.

File: custom_messages/nodes/fiducial_tracker.py
Comment (jim-v) [inline custom_messages/nodes/fiducial_tracker.py]: 2019

File: custom_messages/nodes/fiducial_tracker.py
Comment (jim-v) [inline custom_messages/nodes/fiducial_tracker.py]: This file looks like it has a lot of similarity with the other fiducial tracker.  Could we import that into this file to avoid the duplicated code?   Or could this version replace the original?

File: fiducial_follow/follower_controller.py
Comment (jim-v) [inline fiducial_follow/follower_controller.py]: `rospy.Rate()` takes a frequency in Hz, so if `loop_msec` is in milliseconds, the argument should be `1000.0 / loop_msec`

File: fiducial_follow/nodes/follow.py
Comment (jim-v) [inline fiducial_follow/nodes/follow.py]: This seems a bit hacky, could we just set `target_fiducial` to None?

File: fiducial_follow/nodes/follow.py
Comment (jim-v) [inline fiducial_follow/nodes/follow.py]: Could we have a list for these, instead of lots of variables?

File: fiducial_follow/nodes/follow.py
Comment (jim-v) [inline fiducial_follow/nodes/follow.py]: Could do: ```cmdType, actOnDone, comComment, strParam1, strParam2, numParam1, numParam2 = newCommand```

File: fiducial_follow/nodes/follow.py
Comment (jim-v) [inline fiducial_follow/nodes/follow.py]: It would be easier to read if the code for each of these commands were in its own function.

File: fiducial_follow/nodes/follow.py
Comment (jim-v) [inline fiducial_follow/nodes/follow.py]: You don't need the parens for an if clause in Python

File: custom_messages/nodes/fiducial_tracker.py
Comment (jim-v) [reply custom_messages/nodes/fiducial_tracker.py]: The .pyc file should not be in the repo.

**Review by jim-v — COMMENTED** [review]
Some of my comments, e.g. parentheses in if clauses, apply in multiple places.
I haven't looked at the code in great detail, or try to run this.

File: custom_messages/msg/FollowerCommand.msg
Comment (mjstn) [reply custom_messages/msg/FollowerCommand.msg:6]: The point is it is a generic command block defined by the command type.  If you take a look at the Readme you will gain insight into its need to be non descriptive.

File: custom_messages/nodes/fiducial_tracker.py
Comment (mjstn) [reply custom_messages/nodes/fiducial_tracker.py]: Ok, thanks.

File: custom_messages/nodes/fiducial_tracker.py
Comment (mjstn) [reply custom_messages/nodes/fiducial_tracker.py]: This new follow.py completely removes the need to have what I had earlier called fiducial_tracker.   So there is not going to be a fiducial tracker, only an improvement you see here to our legacy fiducial follow.

File: fiducial_follow/follower_controller.py
Comment (mjstn) [reply fiducial_follow/follower_controller.py]: I tried to make the frequency of the loop exactly like before but only have it be in a variable.

File: fiducial_follow/nodes/follow.py
Comment (mjstn) [reply fiducial_follow/nodes/follow.py]: Yes and in fact the readme says exactly that.   I will change this to say 'none' to be more clear.  It ends up being a string compare anyway so 'none' is ok.

File: fiducial_follow/nodes/follow.py
Comment (mjstn) [reply fiducial_follow/nodes/follow.py]: The functionality in need is to have the string appear in only one place that means a specific command.  Here I was doing the equivalent of a define.  What  I would do in C would use an array of structs and in C++ a map container.    In Python I don't know the ideal approach but I could use a  list of tuples to satisfy your request you make where each tuple is for a specific command which is not elegant but would have minimal risk of munging up the already debugged code and would avoid hard coded strings.  There may be a 'container' in python that can be accessed by a string and if no 'hit' you knew it was not in that container.

File: custom_messages/msg/FollowerCommand.msg
Comment (jim-v) [reply custom_messages/msg/FollowerCommand.msg:6]: That makes sense

File: fiducial_follow/follower_controller.py
Comment (jim-v) [reply fiducial_follow/follower_controller.py]: The units of that variable would be Hz not me.

File: fiducial_follow/nodes/follow.py
Comment (jim-v) [reply fiducial_follow/nodes/follow.py]: I meant the keyword None not the string none, but that would work

Comment (jim-v) [reply fiducial_follow/nodes/follow.py]: I meant the keyword None not the string none, but that would work

Comment (jim-v) [reply fiducial_follow/nodes/follow.py]: I meant the keyword None not the string none, but that would work

File: fiducial_follow/nodes/follow.py
Comment (jim-v) [reply fiducial_follow/nodes/follow.py]: It might be better to have module level variables than class instance variables for this purpose. You could also use an enum or a dictionary instead of individual variables.

Comment (jim-v) [reply fiducial_follow/nodes/follow.py]: Python dictionaries are like C++ maps, but more friendly

File: custom_messages/package.xml
Comment (jim-v) [inline custom_messages/package.xml:16]: Should set the license and a valid maintainer email address at minimum

File: fiducial_follow/README.md
Comment (jim-v) [inline fiducial_follow/README.md]: Confusing sentence.  I read it several times and am still not sure what it means.

File: fiducial_follow/README.md
Comment (jim-v) [inline fiducial_follow/README.md:69]: Could you create a PR to add the launch file to raspicam_node instead of describing how to create it.

**Review by jim-v — APPROVED** [review]

File: custom_messages/package.xml
Comment (mjstn) [reply custom_messages/package.xml:16]: Ok, I just used what was there before but I'll fill it in now.

File: fiducial_follow/README.md
Comment (mjstn) [reply fiducial_follow/README.md]: Ok, I agree, quite a run-on sentence

File: fiducial_follow/README.md
Comment (mjstn) [reply fiducial_follow/README.md:69]: Yes that should happen but it is a chicken and egg thing.  I would first create the raspicam yaml and launch and THEN get back to this to have a launch of  fiducial_floor_follow.launch.
One blocker on that is if a full on calibration has to happen OR can do like I did today (cheap shot) and form the yaml and raspicam launch for 1640x1232 with copies of the 1280x960 files edited.   Does the plum-bob stuff still hold true if I just scale up the pixel count or is new cal required.   In my own environment now I did the copies then edits to form the required 1640x1232 files and it runs but I bet maybe not a good cal.

Comment (jim-v) [reply fiducial_follow/README.md:69]: Yeah, this comes up quite a bit, because we have multiple repos.  What you could do is wait until both PRs are approved and then merge them at the same time.

The camera calibration includes image size, the focal length in both directions, the optical center of the image and the distortion coefficients.  You can change the image size and center for the new size and it will probably work quite well, but it is better to do a calibration with a checkerboard.   I have one you can borrow.

Comment (jim-v) [reply fiducial_follow/README.md:69]: It might be worth committing the dubious calibration now and creating an issue to create a better one.  That way it is documented that it needs to be done.

**Pull request closed without merging**
